stacks-signer/src/crypto/frost.rs

```diff
@@ -0,0 +1,780 @@
+        let sig_share_request = SignatureShareRequest {
+            dkg_id: self.current_dkg_id,
+            sign_id: self.current_sign_id,
+            correlation_id: 0, // TODO: what is this?
+            nonce_responses,
+            message: self.message.clone(),
+        };
+        let sig_share_request_msg = Message {
+            sig: sig_share_request.sign(&self.message_private_key).expect(""),
+            msg: MessageTypes::SignShareRequest(sig_share_request), // TODO: either SigShare or SignatureShare, SignShare is right out
+        };
+        self.ids_to_await = (0..self.total_signers).collect();
+        self.move_to(State::SigShareGather)?;
+        Ok(sig_share_request_msg)
+    }
+
+    fn gather_sig_shares(&mut self, message: &Message) -> Result<(), Error> {
+        match &message.msg {
+            MessageTypes::SignShareResponse(sig_share_response) => {
+                // TODO: check sign_id
+                self.signature_shares
+                    .insert(sig_share_response.signer_id, sig_share_response.signature_shares.clone());
+                self.ids_to_await.remove(&sig_share_response.signer_id);
+                info!(
+                    "Sign round #{} SignShareResponse from signer #{}. Waiting on {:?}",
+                    sig_share_response.sign_id, sig_share_response.signer_id, self.ids_to_await
+                );
+            }
+            _ => {}
+        }
+        if self.ids_to_await.is_empty() {
+            // Calculate the aggregate signature
+            let polys: Vec<PolyCommitment> = self
+                .dkg_public_shares
+                .values()
+                .map(|ps| ps.public_share.clone())
+                .collect();
+
+            let nonce_responses = (0..self.total_signers).map(|i| self.public_nonces[&i].clone()).collect::<Vec<NonceResponse>>();
+
+            let nonces = nonce_responses
+                .iter()
+                .flat_map(|nr| nr.nonces.clone())
+                .collect::<Vec<PublicNonce>>();
+
+            let shares = &self
+                .public_nonces
+                .iter()
+                .flat_map(|(i, _)| self.signature_shares[i].clone())
+                .collect::<Vec<SignatureShare>>();
+
+            info!(
+                "aggregator.sign({:?}, {:?}, {:?})",
+                self.message,
+                nonces.len(),
+                shares.len()
+            );
+
+            let mut aggregator = v1::SignatureAggregator::new(self.total_keys, self.threshold, polys)?;
+
+            let sig = aggregator.sign(&self.message, &nonces, shares)?;
+
+            info!("Signature ({}, {})", sig.R, sig.z);
+
+            let proof = SchnorrProof::new(&sig)?;
+
+            info!("SchnorrProof ({}, {})", proof.r, proof.s);
+
+            if !proof.verify(&self.aggregate_public_key.x(), &self.message) {
+                warn!("SchnorrProof failed to verify!");
```

Comment:
From the demo today, it seemed that this happened?  Can this happen as part of the signer's normal operation?  If so, what would the signer need to do here?

Reply:
We didn't see the actual warning, but we did see multiple DkgPublic and Nonce rounds.  So I think we just ran out of time for the 30 seconds the integration test allocates before stopping the Coordinator.

In a follow on PR I will add another channel so that the Signer can tell us results of operations, like the channel I added so we can give it commands async.  I can do that now if you'd like.

Reply:
Lol actually I had bad copypasta, the aggregate public key was actually being computed not the nonce XD

Fixed now, will test repeatedly until I see this nonce repeat and it still works.

Reply:
Yup, just watched current code do 3 nonce rounds before getting an even Y, and SchnorrProof worked.

Fixed for realz.

Reply:
Follow on branch is ready with extra channel for `OperationResult`s.  Will wait until this is merged and try to batch a bunch of upstream changes.

Reply:
Yeah adding this channel should also make the return time of the run_one_pass irrelevant and is closely related to this issue. https://github.com/stacks-network/stacks-blockchain/issues/3916

Reply:
@jferrant here's a draft of the follow on PR with a results channel: 

https://github.com/stacks-network/stacks-blockchain/pull/3923